### ISSUE TITLE
Implementing gradient scaling with class frequencies on Softmax Loss Layer.

### ISSUE THREAD

**Issue by erogol** [BVLC/caffe#3250]
Given a dataset with imbalanced class distribution learning a robust NN model is hard. In general, the network leads a bias toward the common classes and ignores the rare ones. For solving this issue, I target to rescale gradient signals regarding the class frequencies. I plan to follow these steps for the implementation;

- Compute class frequency coefficients which are less then or equal to 1 and 1 for the smallest class and  min_class_freq/class_freq for the other classes. These values are written to binary proto file as a list.

- I give an extra proto parameter to SoftmaxWithLoss layer as the file name of this freq. file similar to mean file used by the data layer.

- At the runtime, softmax layer read the file and define a blob including these freq values, at the setup.

- Then for each iteration I scale the gradient with corresponding frequency value indexed by the label value.

Here is the basic change on softmax_loss_layer.cu to perform the scaling.

````
template <typename Dtype>
__global__ void SoftmaxLossBackwardGPU1(const int nthreads, const Dtype* top,
          const Dtype* label, Dtype* bottom_diff, const int num, const int dim,
          const int spatial_dim, const bool has_ignore_label_,
          const int ignore_label_, Dtype* counts, Dtype* class_freqs) {
  const int channels = dim / spatial_dim;

  CUDA_KERNEL_LOOP(index, nthreads) {
    const int n = index / spatial_dim;
    const int s = index % spatial_dim;
    const int label_value = static_cast<int>(label[n * spatial_dim + s]);

    if (has_ignore_label_ && label_value == ignore_label_) {
      for (int c = 0; c < channels; ++c) {
        bottom_diff[n * dim + c * spatial_dim + s] = 0;
      }
      counts[index] = 0;
    } else {
      bottom_diff[n * dim + label_value * spatial_dim + s] -= 1;
      bottom_diff[n * dim + label_value * spatial_dim + s] *= class_freqs[label_value];
      counts[index] = 1;
    }
  }
}

template <typename Dtype>
__global__ void SoftmaxLossBackwardGPU2(const int nthreads, const Dtype* top,
          const Dtype* label, Dtype* bottom_diff, const int num, const int dim,
          const int spatial_dim, const bool has_ignore_label_,
          const int ignore_label_, Dtype* counts) {
  const int channels = dim / spatial_dim;

  CUDA_KERNEL_LOOP(index, nthreads) {
    const int n = index / spatial_dim;
    const int s = index % spatial_dim;
    const int label_value = static_cast<int>(label[n * spatial_dim + s]);

    if (has_ignore_label_ && label_value == ignore_label_) {
      for (int c = 0; c < channels; ++c) {
        bottom_diff[n * dim + c * spatial_dim + s] = 0;
      }
      counts[index] = 0;
    } else {
      bottom_diff[n * dim + label_value * spatial_dim + s] -= 1;
      counts[index] = 1;
    }
  }
}

template <typename Dtype>
void SoftmaxWithLossLayer<Dtype>::Backward_gpu(const vector<Blob<Dtype>*>& top,
    const vector<bool>& propagate_down, const vector<Blob<Dtype>*>& bottom) {
  if (propagate_down[1]) {
    LOG(FATAL) << this->type()
               << " Layer cannot backpropagate to label inputs.";
  }
  if (propagate_down[0]) {
    Dtype* bottom_diff = bottom[0]->mutable_gpu_diff();
    const Dtype* prob_data = prob_.gpu_data();
    const Dtype* top_data = top[0]->gpu_data();
    caffe_gpu_memcpy(prob_.count() * sizeof(Dtype), prob_data, bottom_diff);
    const Dtype* label = bottom[1]->gpu_data();
    const int dim = prob_.count() / outer_num_;
    const int nthreads = outer_num_ * inner_num_;
    // Since this memory is never used for anything else,
    // we use to to avoid allocating new GPU memory.
    Dtype* counts = prob_.mutable_gpu_diff();
    // NOLINT_NEXT_LINE(whitespace/operators)
    if (class_freq_norm_){
      Dtype* class_freqs = class_freqs_.mutable_gpu_data();
      SoftmaxLossBackwardGPU1<Dtype><<<CAFFE_GET_BLOCKS(nthreads),
          CAFFE_CUDA_NUM_THREADS>>>(nthreads, top_data, label, bottom_diff,
          outer_num_, dim, inner_num_, has_ignore_label_, ignore_label_, counts, 
          class_freqs);
    }else{
      SoftmaxLossBackwardGPU2<Dtype><<<CAFFE_GET_BLOCKS(nthreads),
          CAFFE_CUDA_NUM_THREADS>>>(nthreads, top_data, label, bottom_diff,
          outer_num_, dim, inner_num_, has_ignore_label_, ignore_label_, counts);
    }
    const Dtype loss_weight = top[0]->cpu_diff()[0];
    if (normalize_) {
      Dtype count;
      caffe_gpu_asum(nthreads, counts, &count);
      caffe_gpu_scal(prob_.count(), loss_weight / count, bottom_diff);
    } else {
      caffe_gpu_scal(prob_.count(), loss_weight / outer_num_, bottom_diff);
    }
  }
}
````

Basically, I decoupled the GPU operations for a normal backpropagation and the scaled one and I scale the gradient value at SoftmaxLossBackwardGPU1.

Even I spend whole my week, I have a problem that cannot be solved. With any setting of learning rate, with this change, network ends up with exploding gradients after some iterations. However, it does not make sense since the frequency values are less then or equal to 1 then we should observe smaller gradients which might cause diminishing gradient but not explosion.

If someone can help me to point out the problem that would be great. I already asked this at the user forum but I could not get ans response. Sorry for the wrong platform for this question.

**Comment by longjon** [comment]
Closing as this does not belong here (as acknowledged by the poster!).

From https://github.com/BVLC/caffe/blob/master/CONTRIBUTING.md:
> _Please do not post usage, installation, or modeling questions, or other requests for help to Issues._
Use the [caffe-users list](https://groups.google.com/forum/#!forum/caffe-users) instead. This helps developers maintain a clear, uncluttered, and efficient view of the state of Caffe.
